pkg/webhooks/prometheusrule/prometheusrule.go

```diff
@@ -0,0 +1,188 @@
+package prometheusrule
+
+import (
+	"fmt"
+	"net/http"
+
+	hookconfig "github.com/openshift/managed-cluster-validating-webhooks/pkg/config"
+	"github.com/openshift/managed-cluster-validating-webhooks/pkg/webhooks/utils"
+
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	admissionctl "sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	WebhookName string = "prometheusrule-validation"
+	docString   string = `Managed OpenShift Customers may not create PrometheusRule in namespaces managed by Red Hat.`
+)
+
+var (
+	timeout        int32 = 2
+	allowedUsers         = []string{"kube:admin", "system:admin", "backplane-cluster-admin", "system:serviceaccount:openshift-monitoring:prometheus-operator"}
+	sreAdminGroups       = []string{"system:serviceaccounts:openshift-backplane-srep"}
+	scope                = admissionregv1.NamespacedScope
+	rules                = []admissionregv1.RuleWithOperations{
+		{
+			Operations: []admissionregv1.OperationType{
+				admissionregv1.Create,
+				admissionregv1.Update,
+				admissionregv1.Delete,
+			},
+			Rule: admissionregv1.Rule{
+				APIGroups:   []string{"monitoring.coreos.com"},
+				APIVersions: []string{"*"},
+				Resources:   []string{"prometheusrules"},
+				Scope:       &scope,
+			},
+		},
+	}
+	log = logf.Log.WithName(WebhookName)
+)
+
+// prometheusruleWebhook validates a prometheusRule change
+type prometheusruleWebhook struct {
+	s runtime.Scheme
+}
+
+// We just need a runtime object to get the namespace
+type prometheusRule struct {
+	runtime.Object
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+}
+
+// NewWebhook creates the new webhook
+func NewWebhook() *prometheusruleWebhook {
+	scheme := runtime.NewScheme()
+	return &prometheusruleWebhook{
+		s: *scheme,
+	}
+}
+
+// Authorized implements Webhook interface
+func (s *prometheusruleWebhook) Authorized(request admissionctl.Request) admissionctl.Response {
+	return s.authorized(request)
+}
+
+func (s *prometheusruleWebhook) authorized(request admissionctl.Request) admissionctl.Response {
+	var ret admissionctl.Response
+
+	pr, err := s.renderPrometheusRule(request)
+	if err != nil {
+		log.Error(err, "Couldn't render a PrometheusRule from the incoming request")
+		return admissionctl.Errored(http.StatusBadRequest, err)
+	}
+
+	if hookconfig.IsPrivilegedNamespace(pr.GetNamespace()) && pr.GetNamespace() != "openshift-user-workload-monitoring" {
```

Comment:
the Privileged namespaces is a long list including all addon namespaces, will this block addon components to create the rule?

Reply:
I have added the changes requested. Can I get LGTM on this? @wanghaoran1988